File: src/pocketledger/sse.clj
(ns pocketledger.sse
  "Datastar SSE response helpers.

   Datastar expects text/event-stream responses with specific event types:
   - datastar-patch-elements  : morph/replace DOM elements
   - datastar-patch-signals   : update reactive signals"
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [hiccup2.core :as h]))

(defn sse-event
  "Build a single SSE event string."
  [event-type data-lines]
  (str "event: " event-type "\n"
       (str/join "\n" (map #(str "data: " %) data-lines))
       "\n\n"))

(defn patch-elements
  "Create a datastar-patch-elements SSE event."
  ([hiccup-fragment]
   (patch-elements {} hiccup-fragment))
  ([{:keys [selector mode]} hiccup-fragment]
   (let [html-str (str (h/html hiccup-fragment))
         lines (cond-> []
                 selector (conj (str "selector " selector))
                 mode     (conj (str "mode " (name mode)))
                 true     (conj (str "elements " html-str)))]
     (sse-event "datastar-patch-elements" lines))))

(defn patch-signals
  "Create a datastar-patch-signals SSE event."
  [signals-map]
  (sse-event "datastar-patch-signals"
             [(str "signals " (json/write-str signals-map))]))

(defn sse-response
  "Return an SSE response with one or more events concatenated."
  [& event-strings]
  {:status 200
   :headers {"Content-Type" "text/event-stream"
             "Cache-Control" "no-cache"
             "Connection" "keep-alive"}
   :body (str/join event-strings)})

(defn parse-signals
  "Extract Datastar signals from the request.
   GET requests send signals as a query param, POST as the body."
  [request]
  (let [raw (or (get-in request [:params :datastar])
                (when-let [body (:body request)]
                  (if (string? body) body (slurp body))))]
    (when (and raw (not (str/blank? raw)))
      (json/read-str raw :key-fn keyword))))

File: ts/src/sse.ts
/**
 * Datastar SSE response helpers.
 *
 * Datastar expects text/event-stream responses with specific event types:
 * - datastar-patch-elements  : morph/replace DOM elements
 * - datastar-patch-signals   : update reactive signals
 */

export function sseEvent(eventType: string, dataLines: string[]): string {
  return `event: ${eventType}\n${dataLines.map(l => `data: ${l}`).join('\n')}\n\n`;
}

export function patchElements(
  html: string,
  options: { selector?: string; mode?: string } = {}
): string {
  const lines: string[] = [];
  if (options.selector) lines.push(`selector ${options.selector}`);
  if (options.mode) lines.push(`mode ${options.mode}`);
  // SSE data lines can't contain newlines - collapse to single line
  lines.push(`elements ${html.replace(/\n\s*/g, '')}`);
  return sseEvent('datastar-patch-elements', lines);
}

export function patchSignals(signals: Record<string, unknown>): string {
  return sseEvent('datastar-patch-signals', [`signals ${JSON.stringify(signals)}`]);
}

export function redirect(url: string): string {
  return sseEvent('datastar-patch-elements', [
    'selector body',
    'mode append',
    `elements <script>setTimeout(() => window.location.href = '${url}')</script>`
  ]);
}

export function sseResponse(...eventStrings: string[]): Response {
  return new Response(eventStrings.join(''), {
    status: 200,
    headers: {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive'
    }
  });
}

export async function parseSignals(request: Request): Promise<Record<string, unknown>> {
  const url = new URL(request.url);
  const datastarParam = url.searchParams.get('datastar');

  if (datastarParam) {
    return JSON.parse(datastarParam);
  }

  if (request.method === 'POST') {
    const body = await request.text();
    if (body && body.trim()) {
      return JSON.parse(body);
    }
  }

  return {};
}
